index/segment/fs/segment.go

```diff
@@ -0,0 +1,477 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sync"
+	"unicode/utf8"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/generated/proto/fswriter"
+	"github.com/m3db/m3ninx/index"
+	"github.com/m3db/m3ninx/index/segment/fs/encoding"
+	"github.com/m3db/m3ninx/postings"
+	"github.com/m3db/m3ninx/postings/pilosa"
+	"github.com/m3db/m3ninx/postings/roaring"
+	"github.com/m3db/m3ninx/x"
+
+	"github.com/couchbase/vellum"
+	vregex "github.com/couchbaselabs/vellum/regexp"
+)
+
+var (
+	errReaderClosed            = errors.New("segment is closed")
+	errUnsupportedMajorVersion = errors.New("unsupported major version")
+	errNotImplemented          = errors.New("operation not implemented")
+
+	minByteKey = []byte{}
+	maxByteKey = []byte(string(utf8.MaxRune))
+)
+
+// NewSegmentOpts represent the collection of parameters to construct a Segment.
+type NewSegmentOpts struct {
```

Comment:
nit: maybe `SegmentData` would be more appropriate since these values aren't actually optional?

Reply:
sure, i'll break into two structs for this.